### PR TITLE
8245053: Keyboard doesn't show when TextInputControl has focus

### PR DESCRIPTION
**Pull request by abhinayagarwal**
In Android, TextInputControls (TextField and TextArea) are responsible for showing and hiding software keyboard. Currently, a focus listener is attached to these controls and is used to toggle the visibility of keyboard. This condition fails in cases where the control already has focus but the keyboard is not visible.

Ideally, the keyboard should be shown again when the user taps on the TextInputControl. 

This PR adds an event handler for `MouseEvent.MOUSE_CLICKED` event and shows the keyboard  if the TextInput control is both editable and focused.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8245053](https://bugs.openjdk.java.net/browse/JDK-8245053): Keyboard doesn't show when TextInputControl has focus 

### Reviewers
 * Johan Vos ([jvos](@johanvos) - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jfx pull/219/head:pull/219`
`$ git checkout pull/219`

### CODE REVIEW

**Comment by kleopatra** [comment]
out off my home-zone here, just fyi and in case it might be relevant to android skins as well: there's an umbrella issue to cleanup skin implementations [JDK-8241364](https://bugs.openjdk.java.net/browse/JDK-8241364) - both the android text skins look like leaking in the manually registered focus listener (not introduced here) and the eventHandler added here. Probably should be removed manually in dispose (or in the case of the focusListener, registered via skin api).

**Comment by abhinayagarwal** [comment]
@kleopatra Done!

File: modules/javafx.controls/src/android/java/javafx/scene/control/skin/TextAreaSkinAndroid.java [hunk 1]
```diff
@@ -26,27 +26,59 @@
...
+     *                                                                         *
+     * Public API                                                              *
+     *                                                                         *
+     **************************************************************************/
+
+    /** {@inheritDoc} */
+    @Override public void dispose() {
+        getSkinnable().removeEventHandler(MouseEvent.MOUSE_CLICKED, mouseEventListener);
+        getSkinnable().focusedProperty().removeListener(weakFocusChangeListener);
+        super.dispose();
```
Comment (kleopatra) [inline modules/javafx.controls/src/android/java/javafx/scene/control/skin/TextAreaSkinAndroid.java:81]: good :) Next step is to implement dispose according to contract (see JDK-8244112). Currently the pattern in dispose is something like:

    if (getSkinnable() == null) return;
    ... do cleanup for this skin
   super.dispose();

**Comment by kleopatra** [comment]
wondering if there is any way to include android specific skins in unit testing? For all (nearly, htmlEditor excluded ;) others we have a test to guard against contract violations (and will soon have tests for memory leaks).

**Comment by abhinayagarwal** [comment]
Indeed. It would be good if we could add a few Android specific tests which are automatically activated when `-PCOMPILE_TARGETS=android` is passed to Gradle.

**Comment by mipastgt** [comment]
Have you considered the case where a TextField is focused by default when its view is shown? As soon as the view is shown the keybord will pop up, even if the user does not plan to edit anything. I just experienced this when I exchanged the Label in the HelloGluon example against a TextField. A lot of strange things than happen. The keyboard pops up, it hides the Gluon nag screen, if you close the keyboard and the nag screen you still see these edit markers and all this just because this text field was focused by default. To my opinion the keyboard and the edit markers should only be shown when the user actually clicks into the text field and thus indicates that he really wants to edit something. I also noticed that after the keyboard was shown I could not normally close the application anymore. (This was on Android with a Galaxy Tab 4)

**Comment by abhinayagarwal** [comment]
@mipastgt The issue happens because of the focus listener. With this PR, we introduce `mouseEventListener` to show the Keyboard on `MOUSE_CLICKED` event. Therefore, the issue brought up by your previous comment should be resolved by updating the `focusChangeListener` to:

```
private final ChangeListener<Boolean> focusChangeListener = (observable, wasFocused, isFocused) -> {
    if (wasFocused && !isFocused) {
        hideSoftwareKeyboard();
    }
};```

We should most probably also introduce an API for showing the keyboard as available on both iOS and Android platforms :)

Thoughts?

**Comment by johanvos** [comment]
It is probably good to add this suggestion in this PR. I'll review again once it is in.

**Pull request closed without merging**